sys/vfs: add vfs_default, configure default fs for same54-xpro

benpicco opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md.
-->

### Contribution description

This adds a `vfs_default` pseudo-module that works analogous to the `netdev_default` module and let's a board specify a default file system and mount point.

As an example I configured littlefs2 for the external flash on the `same54-xpro` which gets mounted to `/nvm`.

### Testing procedure

Add 

    USEMODULE += vfs_default

to any application (e.g. `examples/gnrc_networking`). You now have a file system! (and a `ls` command) 

Add the

    USEMODULE += vfs_auto_format

module if you want to automatically create the file system too. 

### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

chrysn commented on makefiles/pseudomodules.inc.mk:
```diff
@@ -219,6 +219,7 @@ PSEUDOMODULES += sys_bus_%
 PSEUDOMODULES += vdd_lc_filter_%
 PSEUDOMODULES += vfs_auto_format
 PSEUDOMODULES += vfs_auto_mount
+PSEUDOMODULES += vfs_default
```
Suggesting addition (for some reason doing that through GitHub suggestions doesn't work with your PRs right now):

```
## @defgroup pseudomodule_vfs_default vfs_default
## @brief Enable default assignments of a board's devices to VFS mount points
##
## When this module is active, boards with additional flash storage will
## automatically mount (and possibly format, if @ref
## pseudomodule_vfs_auto_format is enabled) their flash devices with a file
## system that is common for that board (or at least common for this board
## within RIOT).
##
## Boards will generally mount to `/` unless they have several storage
## backends.
PSEUDOMODULES += vfs_default
```

... and as I missed the opportunity to document them right in your original PR:

```
## @defgroup pseudomodule_vfs_auto_format vfs_auto_format
## @brief Format mount points at startup unless they can be mounted
##
## When this module is active, mount points configured through the @ref
## pseudomodule_vfs_auto_mount module that can not be mounted at startup are
## formatted and, if that operation is successful, attempted to mount again.
##
## Beware that this may be a harmful procedure in case a bug that corrupts a
## filesystem coincides with a bug that sends the device into a reboot loop.
PSEUDOMODULES += vfs_auto_format

## @defgroup pseudomodule_vfs_auto_mount vfs_auto_mount
## @brief Mount file systems at startup
##
## When this module is active, mount points specified through
## @ref VFS_AUTO_MOUNT are mounted at their designated mount points at startup.
## These mount points can be specified by the application, or are provided by
## some boards if the @ref pseudomodule_vfs_default module is active.
PSEUDOMODULES += vfs_auto_mount
```

benpicco replied:
Thank you!

benpicco commented on sys/Makefile.dep:
```diff
@@ -498,6 +498,10 @@ ifneq (,$(filter devfs,$(USEMODULE)))
 endif
 
+ifneq (,$(filter vfs_default,$(USEMODULE)))
+  USEMODULE += vfs
```
vfs_auto_mount gets added as a default module if vfs is selected.
We could change it to make vfs_default select it instead

chrysn replied:
No strong preference, but it might make sense. Having XFA components active only when there is an active user generally saves a few bytes, and having default-modules pulled in through a module that obviously pulls in things vs. modules that just provide some API and code doesn't sound wrong either..

benpicco replied:
I think it makes sense to use `vfs_default` for selecting the auto-mounter.

benpicco commented:
> Boards will generally mount to `/` unless they have several storage backends.

Actually that's not so nice.
What's also not so nice is that there is no way to iterate over all mount points. I fixed that by adding the virtual `mountfs` which gets mounted to `/` if `mountfs` is used (enabled by `vfs_default`), so `ls /` returns sensible values now:

```
> ls /
const
total 0 files
> mount
/sda successfully mounted
> ls /
sda
const
total 0 files
> ls /foo/bar/
vfs_opendir error: -ENOENT
> ls /const
hello-world	14 B
hello-riot	13 B
total 2 files
```

I changed the default mount point for flash to `/nvm` to avoid conflicts and keep the name short.

chrysn commented:
On Sat, Feb 12, 2022 at 04:00:57PM -0800, benpicco wrote:
> What's also not so nice is that there is no way to iterate over all
> mount points. I fixed that by adding the virtual `mountfs` which gets
> mounted to `/` if `mountfs` is used (enabled by `vfs_default`), so `ls
> /` returns sensible values now:

There is actually an issue that discussed this --
<https://github.com/RIOT-OS/RIOT/issues/15291>. my conclusion back then
was that a mountfs is probably overkill, and applications already know
where to put their files anyway (but a user-visibile tool, maybe even
`ls`, may want to fall back to going through df). 

One reason that seems relevant now that file systems consistently report
only slash-free strings is that mount points may not all be at `/foo`,
so a mountfs would need to synthesize all the different parts (say,
`/mnt` to get to `/mnt/c`), whereas a userspace tool could say "Nothing
found at `/`, but you may wan to look at the individual mount points,
which are `/mnt/c` and `/nvm`). Also, if something is actually mounted
*inside* a mount point (say, mounts at `/` and `/var`), the mountfs
wouldn't expose that, whereas a user-facing tool could advertise that
"in addition, there's also mount points under where you queried". That's
not to say that this is necessarily the right way, just a few
considerations from this having been sitting for half a year.

benpicco commented:
Alright, I move this to it's own PR then as it's unrelated to the question of default mounts.
Mount point names are still an interesting discussion.

chrysn reviewed (approved):
Looks good to me; I trust you tested this on same54 (but sitting on a preestablished flash device, if the one works the rest is just a matter of the automounter working; that is also the one it makes sense to expose, for the other MTD device is just containing the MAC address).

From my PoV this can be merged as is. If you want further input (as indicated in the "Discussion: RFC" label), please consider setting the "> 1 ACK" flag too to prevent any overzealous merges.

benpicco commented:
I think the `vfs_default` module is already a good improvement, we can improve out file system naming convention if they prove to be insufficient.
